packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart

```diff
@@ -0,0 +1,823 @@
+  /// necessary entries for each deferred component.
+  /// Generates correct file if needed.
+  androidRootSetup,
+  /// Checks for the existence of an android dynamic feature
+  /// for each deferred component. Generates correct file if needed.
+  androidComponentSetup,
+  /// Checks for the loading unit -> deferred component
+  /// mapping in the base feature's AndroidManifest.xml and
+  /// component name string resources in strings.xml.
+  /// Generates correct file if needed.
+  androidLoadingUnitComponentMapping,
+  /// Checks if changes were detected in the loading units
+  /// generated vs `deferred_components_golden.yaml`.
+  loadingUnitGolden,
+  /// Writes the latest generated loading units into the
+  /// `deferred_components_golden.yaml`.
+  writeGolden,
+  /// Deletes the temporary directory that recommended file changes
+  /// are generated into.
+  clearTempDir,
+  /// Generates a diff file that may be applied to the `android` directory
+  /// to apply changes recommended by the setup process.
+  generateDiff,
+}
+
+/// A class to configure and run a deferred component setup verification check.
+///
+/// To configure a setup verification run, supply a list of DeferredComponentsSetupValidatorCheck
+/// values to the contstructor. To run the the setup verification, invoke [runSetup] which
+/// will run each check in the order provided. The results of the checks will be displayed.
+///
+/// Failing a check will result in the tool to exit, ending the current command unless false is
+/// passed to the [exitOnFail] parameter.
+class DeferredComponentsSetupValidator {
+  DeferredComponentsSetupValidator(this.checks, this.env, {
+    String title,
+    Logger logger,
+    Directory templatesDir,
+  }) : _outputDir = env.projectDir
+      .childDirectory('build')
+      .childDirectory(kDeferredComponentsTempDirectory),
+      _inputs = <File>[],
+      _outputs = <File>[],
+      _logger = logger ?? env.logger,
+      _title = title ?? 'Deferred components setup verification',
+      _templatesDir = templatesDir;
+
+  final List<DeferredComponentsSetupValidatorCheck> checks;
+
+  final Environment env;
+
+  final String _title;
+
+  final Logger _logger;
+
+  final Directory _templatesDir;
+
+  /// The name of the golden file that tracks the latest loading units generated.
+  @visibleForTesting
+  static const String kDeferredComponentsGoldenFileName = 'deferred_components_golden.yaml';
+  /// The directory in the build folder to generate missing/modified files into.
+  @visibleForTesting
+  static const String kDeferredComponentsTempDirectory = 'android_deferred_components_setup_files';
+
+  final Directory _outputDir;
+
+  final List<File> _inputs;
+  final List<File> _outputs;
+
+  Future<void> runSetup({
+    // only needed if running androidRootSetup, androidComponentSetup, androidStringMapping
+    List<DeferredComponent> components,
+    // only needed if running androidStringMapping, loadingUnitGolden or writeGolden
```

Comment:
These sorts of comments should really be captured in dartdoc if they're non-trivial